pangeo_forge/recipe.py

```diff
@@ -0,0 +1,320 @@
+    #     return _finalize
+
+
+# Notes about dataclasses:
+# - https://www.python.org/dev/peps/pep-0557/#inheritance
+# - https://stackoverflow.com/questions/51575931/class-inheritance-in-python-3-7-dataclasses
+# This means that, for now, I can't get default arguments to work.
+
+
+@dataclass
+class FSSpecFileOpenerMixin:
+    # input_open_kwargs: dict #= field(default_factory=dict)
+
+    @contextmanager
+    def input_opener(self, fname, **kwargs):
+        logger.info(f"Opening input '{fname}'")
+        with fsspec.open(fname, **kwargs) as f:
+            yield f
+
+
+@dataclass
+class InputCachingMixin(FSSpecFileOpenerMixin):
+    require_cache: bool  # = False
+    input_cache: InputCache
+
+    # returns a function that takes one input, the input_key
+    # this allows us to parallelize these operations
+    @property
+    def cache_input(self):
+
+        opener = super().input_opener
+
+        def cache_func(fname: str) -> None:
+            logger.info(f"Caching input '{fname}'")
+            with opener(fname, mode="rb") as source:
+                with self.input_cache.open(fname, mode="wb") as target:
+                    target.write(source.read())
+
+        return cache_func
+
+    @contextmanager
+    def input_opener(self, fname):
+        if self.input_cache.exists(fname):
+            logger.info(f"Input '{fname}' found in cache")
+            with self.input_cache.open(fname, mode="rb") as f:
+                yield f
+        elif self.require_cache:
+            # this creates an error on prepare because nothing is cached
+            raise IOError("Input can only be opened from cache. Call .cache_input first.")
+        else:
+            logger.info(f"Input '{fname}' not found in cache. Opening directly.")
+            # This will bypass the cache. May be slow.
+            with super().input_opener(fname, mode="rb") as f:
+                yield f
+
+
+@dataclass
+class XarrayInputOpener:
+    xarray_open_kwargs: dict
+
+    def open_input(self, fname):
+        with self.input_opener(fname) as f:
+            logger.info(f"Opening input with Xarray '{fname}'")
+            ds = xr.open_dataset(f, **self.xarray_open_kwargs).load()
+            # do we always want to remove encoding? I think so.
+        ds = fix_scalar_attr_encoding(ds)
+        logger.debug(f"{ds}")
+        return ds
+
+
+@dataclass
+class XarrayConcatChunkOpener(XarrayInputOpener):
+    xarray_concat_kwargs: dict
+
+    def open_chunk(self, chunk_key):
+        logger.info(f"Concatenating inputs for chunk '{chunk_key}'")
+        inputs = self.inputs_for_chunk(chunk_key)
+        dsets = [self.open_input(i) for i in inputs]
+        # CONCAT DELETES ENCODING!!!
+        ds = xr.concat(dsets, self.sequence_dim, **self.xarray_concat_kwargs)
+        logger.debug(f"{ds}")
+
+        # do we really want to just delete all encoding?
+        # for v in ds.variables:
+        #    ds[v].encoding = {}
+
+        # TODO: maybe do some chunking here?
```

Comment:
can we make these options? I think that yes, we generally want to remove netcdf encoding before writing to zarr. But there are probably cases where this isn't true.

Reply:
Yes agreed.